### PR TITLE
Inference of GCN model on unseen test data

### LINKED ISSUE #101: [FEATURE REQUEST] Implement inference function
Implement an inference model to run on graphs, and retrieve node and edge classifications/probabilities.

### LINKED ISSUE #133: Graph pre-process
It would be useful to have  a pre-processing module which processes the graph prior to any calculations. Such as : 

 discarding nodes : This can be edge nodes or specific region of the image subject to input from the user.

### LINKED ISSUE #188: Run model inference on random graph + real annotated data
Write a notebook, to start with.

### LINKED ISSUE #189: Transform inference procedure from notebook to script

### LINKED ISSUE #194: Measure area under [receiver-operator, precision-recall] curve metrics
Perform more elaborate GCN evaluation at inference time.

### LINKED ISSUE #221: [DEVELOPMENT] Split the unknown masking for nodes & edges
Use an extra config hyperparameter for keeping the unknown edges but not the nodes in classifier training

### LINKED ISSUE #224: [DEVELOPMENT] Rename `normalise` to `normalize` for tests to pass

### LINKED ISSUE #225: [DEVELOPMENT] Repurpose 2x2 confusion matrix for use outside of object metrics
Currently, the `plot_confusion_matrix` is exclusive for use in `ExactMetricsComputer`. Would be cool to make this accessible to other notebooks / functions, too:

https://github.com/alan-turing-institute/grace/blob/488f60bf6822106129c1de1c83b244c051dfc827/grace/evaluation/metrics.py#L153-L157

### PR DESCRIPTION
**Pull request by KristinaUlicna**
## PR contribution summary

**Why is this PR useful / good for?** *Please describe the problem(s) you're trying to address.*
* Assuming we have a working GCN, this PR provides a new notebook to run inference & evaluate the performance of the GCN on real, annotated data. 
* It builds a `GraphLabelPredictor` which updates the `GraphAttrs.XXX_PREDICTION` by a vector of numbers for each node and edge: 
https://github.com/alan-turing-institute/grace/blob/1e60b41acbe24d28c6bf68f005e9f26405182adb/grace/evaluation/inference.py#L35-L41
* It allows node `Annotation.UNKNOWN` relabelling to be separated for nodes & edges.
* This PR additionally annotates other notebooks to make sure that the flow of what is done is clearly explained.

### List of proposed changes / linked issues & discussions

- [x] Resolves #221
- [x] Resolves #224
- [x] Resolves #194 
- [x] Resolves #188 
- [x] Resolves #189  
- [x] Resolves #101
- [x] Resolved #133 
- [x] Resolves #225

### What should a reviewer concentrate their feedback on?

+ ✅ Scripts to check
+ 🏃 Notebooks to run (especially `infer_predictions.ipynb`, others only have minor documentation / comments changes)
+ 💻 Code quality
+ 📝 Everything looks OK?

#### What type of PR is this? (check all applicable)

+ 🪄 Feature
+ 🐛 Bug fix
+ #️⃣ Documentation / code annotation
+ 🔥 Performance Improvements

#### Added tests?

+ 🙋 no, because I need some help

---

## PR review summary

Describe what this PR does & **how you reviewed** the individual items, where needed:

* <TEXT>

*Some helper checks to tick off:*

* Focus on image annotation
* Focus on model training
* Could any optimization be applied?
* Is there any redundant code?
* Are there any spelling errors?

**In conclusion**, after my review, I'd like to:

* 🙋 ask some clarifying questions
* 🙅 suggest some specific changes

### CODE REVIEW

File: notebooks/infer_predictions.ipynb [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from grace.io.image_dataset import ImageGraphDataset\n",
+    "from grace.models.feature_extractor import FeatureExtractor\n",
+    "from grace.evaluation.visualisation import plot_simple_graph\n",
```
Comment (crangelsmith) [inline notebooks/infer_predictions.ipynb:29]: I get an error trying to run this cell:
```ModuleNotFoundError: No module named 'grace.evaluation'``` 

I think there is a missing __init__.py in the evaluation directory that would make this work.

File: notebooks/infer_predictions.ipynb [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    "### Read some real grace-annotated data:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "extractor_filename = \"/Users/kulicna/Desktop/classifier/extractor/resnet152.pt\"\n",
```
Comment (crangelsmith) [inline notebooks/infer_predictions.ipynb:46]: If we want people to be able to run this easily maybe we need to add the line to download the resnet before or add a comment to point to where this is.

File: notebooks/infer_predictions.ipynb [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    "feature_extractor = FeatureExtractor(model=pre_trained_resnet)\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "grace_path = \"/Users/kulicna/Desktop/dataset/shape_stars/train\"\n",
```
Comment (crangelsmith) [inline notebooks/infer_predictions.ipynb:57]: In general we should remove path to personal directories if we want these notebooks to be run easily for everyone, we could add a new variable called: `input_paht = /path/to/data/` to be replaced by the user. 

Also, we need to put this data somewhere and easy to download.

**Comment by crangelsmith** [comment]
Hi @KristinaUlicna, the notebook runs except for the minor comments above but I'm a bit confused about why is the inference done on a notebook and not added to the run.py script as part of the pipeline. Couldn't we just add an `eval` option to the config file and do this in run.py after training?

**Comment by crangelsmith** [comment]
Also, it would be good to visualise the resulting graph on inference, as described in https://github.com/alan-turing-institute/grace/issues/222

File: notebooks/infer_predictions.ipynb (hunk 1 above)
Comment (KristinaUlicna) [reply notebooks/infer_predictions.ipynb:29]: Adding the `__init__.py` in the next commit 🚀

File: notebooks/infer_predictions.ipynb (hunk 2 above)
Comment (KristinaUlicna) [reply notebooks/infer_predictions.ipynb:46]: Great point! I included the instructions on how to download the ResNet locally in the `training` subfolder README:
https://github.com/alan-turing-institute/grace/blob/488f60bf6822106129c1de1c83b244c051dfc827/grace/training/README.md?plain=1#L67-L78
Would you like to see it added in the notebook, too?

File: notebooks/infer_predictions.ipynb (hunk 3 above)
Comment (KristinaUlicna) [reply notebooks/infer_predictions.ipynb:57]: Fully agreed, we had this problem in the `examples\` folder where I replaced the path for a user-specified input: 
https://github.com/alan-turing-institute/grace/blob/488f60bf6822106129c1de1c83b244c051dfc827/examples/show_data.py#L13-L20
I'm not sure how to do this from the notebook tho...

**Comment by KristinaUlicna** [comment]
> Also, it would be good to visualise the resulting graph on inference, as described in #222

Coming up in PR #229 ! 🚀

**Review by crangelsmith — APPROVED** [review]

**Pull request merged**